cpp/src/groupby/hash/groupby.cu

```diff
@@ -421,36 +425,31 @@ void sparse_to_dense_results(table_view const& keys,
  * @brief Construct hash map that uses row comparator and row hasher on
  * `d_keys` table and stores indices
  */
-auto create_hash_map(table_device_view const& d_keys,
-                     bool keys_have_nulls,
-                     null_policy include_null_keys,
-                     rmm::cuda_stream_view stream)
+auto create_hash_map(
+  std::shared_ptr<cudf::experimental::row::hash::preprocessed_table> preprocessed_keys,
+  cudf::size_type const num_keys,
+  null_policy const include_null_keys,
+  rmm::cuda_stream_view stream)
 {
   size_type constexpr unused_key{std::numeric_limits<size_type>::max()};
   size_type constexpr unused_value{std::numeric_limits<size_type>::max()};
 
-  using map_type =
-    concurrent_unordered_map<size_type,
-                             size_type,
-                             row_hasher<cudf::detail::default_hash, nullate::DYNAMIC>,
-                             row_equality_comparator<nullate::DYNAMIC>>;
-
   using allocator_type = typename map_type::allocator_type;
 
   auto const null_keys_are_equal =
     include_null_keys == null_policy::INCLUDE ? null_equality::EQUAL : null_equality::UNEQUAL;
+  auto const has_null = nullate::DYNAMIC{true};  // always has null TODO: why??
 
-  row_hasher<cudf::detail::default_hash, nullate::DYNAMIC> hasher{nullate::DYNAMIC{keys_have_nulls},
-                                                                  d_keys};
-  row_equality_comparator rows_equal{
-    nullate::DYNAMIC{keys_have_nulls}, d_keys, d_keys, null_keys_are_equal};
+  cudf::experimental::row::equality::self_comparator row_equal(preprocessed_keys);
```

Comment:
This could be part of your mysterious bug above, but this has a lifetime problem. The `self_comparator` object needs to stay alive as long as the hash map does. Otherwise the `key_equal` returned from `device_comparator` has potentially dangling references. 

The model for the new comparators is that they need to outlive any of the non-owning references to them returned from the `device_comparator`.

Reply:
Done.